File: Day22/Part2/make_array_continuous.cpp
#include <bits/stdc++.h>
using namespace std;

// LC 2009 — Minimum Number of Operations to Make Array Continuous
// Sort unique + two pointers: maximize count kept within window [x, x+n-1]

int minOperations(vector<int> nums){
    int n = nums.size();
    sort(nums.begin(), nums.end());
    nums.erase(unique(nums.begin(), nums.end()), nums.end());
    int m = nums.size();
    int ansKeep = 0;
    int j = 0;
    for(int i=0;i<m;++i){
        while(j<m && nums[j] < nums[i] + n) ++j;
        ansKeep = max(ansKeep, j - i);
    }
    return n - ansKeep;
}

int main(){
    cout << minOperations({4,2,5,3}) << "\n"; // 0
    cout << minOperations({1,2,3,5,6}) << "\n"; // 1
    return 0;
}

File: learnLeetCode/Day22/Part2/make_array_continuous.py
# LC 2009 — Minimum Number of Operations to Make Array Continuous
# Sort unique + sliding window to keep most that already fit into a window of size n (values within [x, x+n-1])
from typing import List

def minOperations(nums: List[int]) -> int:
    n = len(nums)
    a = sorted(set(nums))
    m = len(a)
    ans_keep = 0
    j = 0
    for i in range(m):
        while j < m and a[j] < a[i] + n:
            j += 1
        ans_keep = max(ans_keep, j - i)
    return n - ans_keep

if __name__ == "__main__":
    print(minOperations([4,2,5,3]))  # 0
    print(minOperations([1,2,3,5,6]))  # 1
